### ISSUE TITLE
Thread-safety of parser, grammar objects

### ISSUE THREAD

**Issue by codeape2** [davidnemeti/Sarcasm#26]
In the code below, could `grammar` and `parser` be shared, reused and potentially used by multiple threads simultaneously?

```
EquationExpression ParseEquation(string s) {
    var grammar = new EquationGrammar();
    var parser = MultiParser.Create(grammar);
    var parseTree = parser.Parse(s);
    if (parseTree.HasErrors()) {
        throw ...
    }
    return parseTree.RootAstValue;
}
```

**Comment by davidnemeti** [comment]
An Irony's **parser** instance - of which Sarcasm is based on - **cannot be used concurrently**.

**Grammar** could be used concurrently on Irony's side, however Sarcasm uses it so due to technical issues that the same grammar instance **cannot be used concurrently**.

**Comment by codeape2** [comment]
Great, thanks for the info!

**Comment by codeape2** [comment]
... and just to be clear: Both a the `grammar` and `parser` instances can be reused, i.e. I can parse a number of strings using the same instances (just not concurrently), right?

**Comment by davidnemeti** [comment]
Yes, both the grammar and the parser can be reused, so you can use the same instance sequentially.
